File: src/main/java/rocks/sakira/sakurarosea/common/item/Items.java
package rocks.sakira.sakurarosea.common.item;

import net.minecraft.item.*;
import net.minecraftforge.fml.DistExecutor;
import net.minecraftforge.fml.RegistryObject;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;
import rocks.sakira.sakurarosea.common.block.Blocks;
import rocks.sakira.sakurarosea.common.tileentities.renderers.SakuraChestItemStackRenderer;
import rocks.sakira.sakurarosea.common.tileentities.renderers.SakuraShieldRenderer;

import static rocks.sakira.sakurarosea.Constants.MOD_ID;

public class Items {
    public static final DeferredRegister<Item> REGISTER = new DeferredRegister<>(ForgeRegistries.ITEMS, MOD_ID);
    private static final ItemGroup GROUP = new SakuraRoseaItemGroup("sakurarosea") {
        @Override
        public ItemStack createIcon() {
            return new ItemStack(SAKURA_SAPLING_BLOCK_ITEM.get());
        }
    };
    public static final RegistryObject<Item> CHERRY_ITEM = REGISTER.register(
            "cherry",

            () -> new Item(
                    new Item.Properties().group(GROUP)
                            .food(Foods.APPLE)
            )
    );

    public static final RegistryObject<Item> SAKURA_BUTTON_ITEM = REGISTER.register(
            "sakura_button",

            () -> new BlockItem(
                    Blocks.SAKURA_BUTTON_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> PINK_CLAY_DOOR_ITEM = REGISTER.register(
            "pink_clay_door",

            () -> new BlockItem(
                    Blocks.PINK_CLAY_DOOR_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> WHItE_CLAY_DOOR_ITEM = REGISTER.register(
            "white_clay_door",

            () -> new BlockItem(
                    Blocks.WHITE_CLAY_DOOR_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_DOOR_ITEM = REGISTER.register(
            "sakura_door",

            () -> new BlockItem(
                    Blocks.SAKURA_DOOR_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_TRAPDOOR_ITEM = REGISTER.register(
            "sakura_trapdoor",

            () -> new BlockItem(
                    Blocks.SAKURA_TRAPDOOR_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_LEAVES_BLOCK_ITEM = REGISTER.register(
            "sakura_leaves",

            () -> new BlockItem(
                    Blocks.SAKURA_LEAVES_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> ALT_SAKURA_LEAVES_BLOCK_ITEM = REGISTER.register(
            "alt_sakura_leaves",

            () -> new BlockItem(
                    Blocks.ALT_SAKURA_LEAVES_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> WHITE_SAKURA_LEAVES_BLOCK_ITEM = REGISTER.register(
            "white_sakura_leaves",

            () -> new BlockItem(
                    Blocks.WHITE_SAKURA_LEAVES_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_LOG_BLOCK_ITEM = REGISTER.register(
            "sakura_log",

            () -> new BlockItem(
                    Blocks.SAKURA_LOG_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_PRESSURE_PLATE_BLOCK_ITEM = REGISTER.register(
            "sakura_pressure_plate",

            () -> new BlockItem(
                    Blocks.SAKURA_PRESSURE_PLATE_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_SAPLING_BLOCK_ITEM = REGISTER.register(
            "sakura_sapling",

            () -> new BlockItem(
                    Blocks.SAKURA_SAPLING_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_SLAB_BLOCK_ITEM = REGISTER.register(
            "sakura_slab",

            () -> new BlockItem(
                    Blocks.SAKURA_SLAB_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_PLANKS_BLOCK_ITEM = REGISTER.register(
            "sakura_planks",

            () -> new BlockItem(
                    Blocks.SAKURA_PLANKS_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> STRIPPED_SAKURA_LOG_BLOCK_ITEM = REGISTER.register(
            "stripped_sakura_log",

            () -> new BlockItem(
                    Blocks.STRIPPED_SAKURA_LOG_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_WOOD_BLOCK_ITEM = REGISTER.register(
            "sakura_wood",

            () -> new BlockItem(
                    Blocks.SAKURA_WOOD_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> STRIPPED_SAKURA_WOOD_BLOCK_ITEM = REGISTER.register(
            "stripped_sakura_wood",

            () -> new BlockItem(
                    Blocks.STRIPPED_SAKURA_WOOD_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> STRIPPED_STAIRS_BLOCK_ITEM = REGISTER.register(
            "sakura_stairs",

            () -> new BlockItem(
                    Blocks.SAKURA_STAIRS_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_FENCE_ITEM = REGISTER.register(
            "sakura_fence",

            () -> new BlockItem(
                    Blocks.SAKURA_FENCE_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_BARREL_ITEM = REGISTER.register(
            "sakura_barrel",

            () -> new BlockItem(
                    Blocks.SAKURA_BARREL_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_CHEST_ITEM = REGISTER.register(
            "sakura_chest",

            () -> new BlockItem(
                    Blocks.SAKURA_CHEST_BLOCK.get().getBlock(),
                    getPropsForChest()
            )
    );

    public static final RegistryObject<Item> SAKURA_FENCE_GATE_ITEM = REGISTER.register(
            "sakura_fence_gate",

            () -> new BlockItem(
                    Blocks.SAKURA_FENCE_GATE_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_SIGN_ITEM = REGISTER.register(
            "sakura_sign",

            () -> new SignItem(
                    new Item.Properties().group(GROUP)
                            .maxStackSize(16),

                    Blocks.SAKURA_SIGN_BLOCK.get(),
                    Blocks.SAKURA_WALL_SIGN_BLOCK.get()
            )
    );

    public static final RegistryObject<Item> SAKURA_BOAT_ITEM = REGISTER.register(
            "sakura_boat",

            () -> new SakuraBoatItem(
                    new Item.Properties().group(GROUP).maxStackSize(1)
            )
    );

    public static final RegistryObject<Item> SAKURA_BLOSSOM_ITEM = REGISTER.register(
            "sakura_blossom",

            () -> new Item(
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_TEA_ITEM = REGISTER.register(
            "sakura_tea",

            () -> new SakuraTeaItem(
                    new Item.Properties().group(GROUP)
                            .maxStackSize(1)
                            .food(Foods.BAKED_POTATO)  // Same hunger/sat that we want
            )
    );

    public static final RegistryObject<Item> SAKURA_SHIELD_ITEM = REGISTER.register(
            "sakura_shield",

            () -> new ShieldItem(getPropsForShield())
    );

    public static final RegistryObject<Item> WHITE_CLAY_ITEM = REGISTER.register(
            "white_clay",

            () -> new Item(
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> WHITE_BRICK_ITEM = REGISTER.register(
            "white_brick",

            () -> new Item(
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> PINK_CLAY_ITEM = REGISTER.register(
            "pink_clay",

            () -> new Item(
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> PINK_BRICK_ITEM = REGISTER.register(
            "pink_brick",

            () -> new Item(
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> WHITE_BRICKS_ITEM = REGISTER.register(
            "white_bricks",

            () -> new BlockItem(
                    Blocks.WHITE_BRICKS_BLOCK.get(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> WHITE_BRICKS_BLOCK_ITEM = REGISTER.register(
            "white_bricks_stairs",

            () -> new BlockItem(
                    Blocks.WHITE_BRICKS_STAIRS_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> WHITE_BRICKS_WALL_ITEM = REGISTER.register(
            "white_bricks_wall",

            () -> new BlockItem(
                    Blocks.WHITE_BRICKS_WALL_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> WHITE_BRICKS_SLAB_ITEM = REGISTER.register(
            "white_bricks_slab",

            () -> new BlockItem(
                    Blocks.WHITE_BRICKS_SLAB_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> PINK_BRICKS_ITEM = REGISTER.register(
            "pink_bricks",

            () -> new BlockItem(
                    Blocks.PINK_BRICKS_BLOCK.get(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> PINK_BRICKS_BLOCK_ITEM = REGISTER.register(
            "pink_bricks_stairs",

            () -> new BlockItem(
                    Blocks.PINK_BRICKS_STAIRS_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> PINK_BRICKS_WALL_ITEM = REGISTER.register(
            "pink_bricks_wall",

            () -> new BlockItem(
                    Blocks.PINK_BRICKS_WALL_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> PINK_BRICKS_SLAB_ITEM = REGISTER.register(
            "pink_bricks_slab",

            () -> new BlockItem(
                    Blocks.PINK_BRICKS_SLAB_BLOCK.get().getBlock(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> SAKURA_LEAF_CARPET_ITEM = REGISTER.register(
            "sakura_leaf_carpet",

            () -> new BlockItem(
                    Blocks.SAKURA_LEAF_CARPET.get(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> ALT_SAKURA_LEAF_CARPET_ITEM = REGISTER.register(
            "alt_sakura_leaf_carpet",

            () -> new BlockItem(
                    Blocks.ALT_SAKURA_LEAF_CARPET.get(),
                    new Item.Properties().group(GROUP)
            )
    );

    public static final RegistryObject<Item> WHITE_SAKURA_LEAF_CARPET_ITEM = REGISTER.register(
            "white_sakura_leaf_carpet",

            () -> new BlockItem(
                    Blocks.WHITE_SAKURA_LEAF_CARPET.get(),
                    new Item.Properties().group(GROUP)
            )
    );

// TODO: Cherry pie textures
//
//    private static final RegistryObject<Item> CHERRY_PIE_ITEM = REGISTER.register(
//            "cherry_pie",
//
//            () -> new BlockItem(
//                    Blocks.CHERRY_PIE_BLOCK.get(),
//                    new Item.Properties().group(GROUP)
//            )
//    );

    private static Item.Properties getPropsForShield() {
        Item.Properties props = new Item.Properties().
                maxDamage(336)
                .group(GROUP);

        DistExecutor.runForDist(
                () -> () -> props.setISTER(() -> SakuraShieldRenderer::new),
                () -> () -> props
        );

        return props;
    }

    private static Item.Properties getPropsForChest() {
        Item.Properties props = new Item.Properties().group(GROUP);

        DistExecutor.runForDist(
                () -> () -> props.setISTER(() -> SakuraChestItemStackRenderer::new),
                () -> () -> props
        );

        return props;
    }
}

File: src/main/kotlin/io/github/faecraft/sakurarosea/register/Items.kt
package io.github.faecraft.sakurarosea.register

import net.fabricmc.fabric.api.client.itemgroup.FabricItemGroupBuilder
import net.fabricmc.fabric.api.item.v1.FabricItemSettings
import net.minecraft.block.Block
import net.minecraft.block.SignBlock
import net.minecraft.block.WallSignBlock
import net.minecraft.entity.vehicle.BoatEntity
import net.minecraft.item.*
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry
import net.minecraft.world.World
import io.github.faecraft.sakurarosea.MOD_ID
import io.github.faecraft.sakurarosea.item.SakuraBoatItem
import io.github.faecraft.sakurarosea.item.SakuraTeaItem
import io.github.faecraft.sakurarosea.register.Entities.sakuraBoatFactory

object Items {
    val GROUP: ItemGroup = FabricItemGroupBuilder
        .create(Identifier(MOD_ID, "sakurarosea"))
        .icon { ItemStack(SAKURA_SAPLING) }
        .build()

    val CHERRY = createFoodItem(FoodComponents.APPLE)
    val CHERRY_PIE = createBlockItem(Blocks.CHERRY_PIE)

    val SAKURA_BLOSSOM = createItem()
    val SAKURA_SAPLING = createBlockItem(Blocks.SAKURA_SAPLING)

    val SAKURA_BUTTON = createBlockItem(Blocks.SAKURA_BUTTON)
    val SAKURA_PRESSURE_PLATE = createBlockItem(Blocks.SAKURA_PRESSURE_PLATE)

    val SAKURA_CRAFTING_TABLE = createBlockItem(Blocks.SAKURA_CRAFTING_TABLE)

    val SAKURA_LEAVES = createBlockItem(Blocks.SAKURA_LEAVES)
    val SAKURA_LEAVES_ALT = createBlockItem(Blocks.SAKURA_LEAVES_ALT)
    val SAKURA_LEAVES_WHITE = createBlockItem(Blocks.SAKURA_LEAVES_WHITE)

    val SAKURA_LEAF_CARPET = createBlockItem(Blocks.SAKURA_LEAF_CARPET)
    val SAKURA_LEAF_CARPET_ALT = createBlockItem(Blocks.SAKURA_LEAF_CARPET_ALT)
    val SAKURA_LEAF_CARPET_WHITE = createBlockItem(Blocks.SAKURA_LEAF_CARPET_WHITE)

    val SAKURA_LOG = createBlockItem(Blocks.SAKURA_LOG)
    val SAKURA_LOG_STRIPPED = createBlockItem(Blocks.SAKURA_LOG_STRIPPED)
    val SAKURA_WOOD = createBlockItem(Blocks.SAKURA_WOOD)
    val SAKURA_WOOD_STRIPPED = createBlockItem(Blocks.SAKURA_WOOD_STRIPPED)

    val SAKURA_PLANKS = createBlockItem(Blocks.SAKURA_PLANKS)

    val SAKURA_BARREL = createBlockItem(Blocks.SAKURA_BARREL)
    val SAKURA_CHEST = createBlockItem(Blocks.SAKURA_CHEST)

    val SAKURA_FENCE = createBlockItem(Blocks.SAKURA_FENCE)
    val SAKURA_FENCE_GATE = createBlockItem(Blocks.SAKURA_FENCE_GATE)
    val SAKURA_SLAB = createBlockItem(Blocks.SAKURA_SLAB)
    val SAKURA_STAIRS = createBlockItem(Blocks.SAKURA_STAIRS)

    val SAKURA_SIGN = createSignItem(Blocks.SAKURA_SIGN, Blocks.SAKURA_SIGN_WALL)

    val SAKURA_DOOR = createBlockItem(Blocks.SAKURA_DOOR)
    val SAKURA_TRAPDOOR = createBlockItem(Blocks.SAKURA_TRAPDOOR)

    val PINK_BRICK = createItem()
    val PINK_CLAY = createItem()
    val PINK_CLAY_DOOR = createBlockItem(Blocks.PINK_CLAY_DOOR)

    val WHITE_BRICK = createItem()
    val WHITE_CLAY = createItem()
    val WHITE_CLAY_DOOR = createBlockItem(Blocks.WHITE_CLAY_DOOR)

    val PINK_BRICKS = createBlockItem(Blocks.PINK_BRICKS)
    val PINK_BRICKS_STAIRS = createBlockItem(Blocks.PINK_BRICKS_STAIRS)
    val PINK_BRICKS_WALL = createBlockItem(Blocks.PINK_BRICKS_WALL)
    val PINK_BRICKS_SLAB = createBlockItem(Blocks.PINK_BRICKS_SLAB)

    val WHITE_BRICKS = createBlockItem(Blocks.WHITE_BRICKS)
    val WHITE_BRICKS_STAIRS = createBlockItem(Blocks.WHITE_BRICKS_STAIRS)
    val WHITE_BRICKS_WALL = createBlockItem(Blocks.WHITE_BRICKS_WALL)
    val WHITE_BRICKS_SLAB = createBlockItem(Blocks.WHITE_BRICKS_SLAB)

//    Fabric Shield Lib has no docs or usable examples, so we're just not going to have a shield for now
//    val SAKURA_SHIELD = createShield()

    val SAKURA_TEA = SakuraTeaItem(
        FabricItemSettings().group(GROUP)
            .maxCount(1)
            .food(FoodComponents.BAKED_POTATO)
    )

    val SAKURA_BOAT = createBoat(::sakuraBoatFactory)

    fun register() {
        register(CHERRY, "cherry")
        register(CHERRY_PIE, "cherry_pie")

        register(SAKURA_BLOSSOM, "sakura_blossom")
        register(SAKURA_SAPLING, "sakura_sapling")

        register(SAKURA_BUTTON, "sakura_button")
        register(SAKURA_PRESSURE_PLATE, "sakura_pressure_plate")

        register(SAKURA_CRAFTING_TABLE, "sakura_crafting_table")

        register(SAKURA_LEAVES, "sakura_leaves")
        register(SAKURA_LEAVES_ALT, "alt_sakura_leaves")
        register(SAKURA_LEAVES_WHITE, "white_sakura_leaves")

        register(SAKURA_LEAF_CARPET, "sakura_leaf_carpet")
        register(SAKURA_LEAF_CARPET_ALT, "alt_sakura_leaf_carpet")
        register(SAKURA_LEAF_CARPET_WHITE, "white_sakura_leaf_carpet")

        register(SAKURA_LOG, "sakura_log")
        register(SAKURA_LOG_STRIPPED, "stripped_sakura_log")
        register(SAKURA_WOOD, "sakura_wood")
        register(SAKURA_WOOD_STRIPPED, "stripped_sakura_wood")

        register(SAKURA_PLANKS, "sakura_planks")

        register(SAKURA_BARREL, "sakura_barrel")
        register(SAKURA_CHEST, "sakura_chest")

        register(SAKURA_FENCE, "sakura_fence")
        register(SAKURA_FENCE_GATE, "sakura_fence_gate")
        register(SAKURA_SLAB, "sakura_slab")
        register(SAKURA_STAIRS, "sakura_stairs")

        register(SAKURA_SIGN, "sakura_sign")

        register(SAKURA_DOOR, "sakura_door")
        register(SAKURA_TRAPDOOR, "sakura_trapdoor")

        register(PINK_BRICK, "pink_brick")
        register(PINK_CLAY, "pink_clay")
        register(PINK_CLAY_DOOR, "pink_clay_door")

        register(WHITE_BRICK, "white_brick")
        register(WHITE_CLAY, "white_clay")
        register(WHITE_CLAY_DOOR, "white_clay_door")

        register(PINK_BRICKS, "pink_bricks")
        register(PINK_BRICKS_STAIRS, "pink_bricks_stairs")
        register(PINK_BRICKS_WALL, "pink_bricks_wall")
        register(PINK_BRICKS_SLAB, "pink_bricks_slab")

        register(WHITE_BRICKS, "white_bricks")
        register(WHITE_BRICKS_STAIRS, "white_bricks_stairs")
        register(WHITE_BRICKS_WALL, "white_bricks_wall")
        register(WHITE_BRICKS_SLAB, "white_bricks_slab")

//        register(SAKURA_SHIELD, "sakura_shield")

        register(SAKURA_TEA, "sakura_tea")

        register(SAKURA_BOAT, "sakura_boat")
    }

    private fun register(item: Item, id: String) {
        Registry.register(Registry.ITEM, Identifier(MOD_ID, id), item)
    }

    // region: Internal builder functions

    private fun createBoat(factory: (World, Double, Double, Double, Float) -> BoatEntity) =
        SakuraBoatItem(factory, FabricItemSettings().group(GROUP))

    private fun createBlockItem(block: Block) = BlockItem(
        block,

        FabricItemSettings().group(GROUP)
    )

    private fun createBlockItem(block: Block, maxCount: Int) = BlockItem(
        block,

        FabricItemSettings()
            .group(GROUP)
            .maxCount(maxCount)
    )

    private fun createFoodItem(food: FoodComponent) = Item(
        FabricItemSettings()
            .group(GROUP)
            .food(food)
    )

    private fun createItem() = Item(FabricItemSettings().group(GROUP))

//    private fun createShield() = BasicShield(
//        FabricItemSettings().group(GROUP).maxCount(1),
//        100,
//        336,
//        SAKURA_PLANKS
//    )

    private fun createSignItem(standingBlock: SignBlock, wallBlock: WallSignBlock) = SignItem(
        FabricItemSettings()
            .group(GROUP)
            .maxCount(16),
        standingBlock,
        wallBlock
    )

    // endregion
}
